plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+    using their metadata for existing saved :class:`~pulpcore.plugin.models.Artifact` objects inside
+    Pulp with the same digest value(s). Any existing :class:`~pulpcore.plugin.models.Artifact`
+    objects found will replace their unsaved counterpart in the
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object.
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): The queue to receive
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        out_q (:class:`asyncio.Queue`): The queue to put
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+    Returns:
+        The query_existing_artifacts stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content = []
+    shutdown = False
+    while True:
+        try:
+            content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content:
+                content = await in_q.get()
+                declarative_content.append(content)
+                continue
+        else:
+            declarative_content.append(content)
+            continue
+
+        all_artifacts_q = Q(pk=None)
+        for content in declarative_content:
+            if content is None:
+                shutdown = True
+                continue
+
+            for declarative_artifact in content.d_artifacts:
+                one_artifact_q = Q()
+                for digest_name in declarative_artifact.artifact.DIGEST_FIELDS:
+                    digest_value = getattr(declarative_artifact.artifact, digest_name)
+                    if digest_value:
+                        key = {digest_name: digest_value}
+                        one_artifact_q &= Q(**key)
+                all_artifacts_q |= one_artifact_q
+
+        for artifact in Artifact.objects.filter(all_artifacts_q):
+            for content in declarative_content:
+                if content is None:
+                    continue
+                for declarative_artifact in content.d_artifacts:
+                    for digest_name in artifact.DIGEST_FIELDS:
+                        digest_value = getattr(declarative_artifact.artifact, digest_name)
+                        if digest_value and digest_value == getattr(artifact, digest_name):
+                            declarative_artifact.artifact = artifact
+        for content in declarative_content:
+            if content is None:
+                continue
+            await out_q.put(content)
+        declarative_content = []
+        if shutdown:
+            break
+    await out_q.put(None)
+
+
+class artifact_downloader:
```

Comment:
Class name should be CapWords - PEP8.

[1] https://www.python.org/dev/peps/pep-0008/#class-names

Reply:
This is going to make some stages lowercase function names and some CapWords. That would be like:

```
stages = [
    self.first_stage.gen_declarative_content,
    query_existing_artifacts, ArtifactDownloader().stage, artifact_saver, 
    query_existing_content_units, content_unit_saver,
    ContentUnitAssociation(new_version).stage
]
```

Do you think we should stick to PEP8 in this case still?

Reply:
Yes.

Reply:
ok I'll port this w/ my next push.

Reply:
@bmbouter you could, in theory, make your classes Callable, so that they can be used as functions

no comment on whether you *should* :)